text/0000-derefered_composite_types.md

```diff
@@ -0,0 +1,110 @@
+- Feature Name: `derefered_composite_types`
+- Start Date: 2023-04-06
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+The proposal of Derefered Composite Types is a Rust unification of Refered Types and Derefered Types.
+
+Derefered Composite Types could help also to get rid of constructor boilerplate.
+
+The symbol `*` before the type name is a marker of Derefered Composite Type. 
+
+The new operator `=&` (and `=&&`, `=&&&`, ...) as a compound of borrow and assignment (`x = &2` same as `x =& 2`) is also required.
+
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Rust has 
+- (1) derefered primitive types (like `i32`)
+- (2) refered primitive types (like `&i32`) 
+- (3) refered composite types (like `Box<i32>`)
+
+But there are no any derefered composite types.
+
+This is not universal. We wish to improve this. So we propose types like `*Box<i32>`.
```

Comment:
This… feels extremely insufficient for motivation. Like, after reading this whole RFC, I mostly get what the feature is, but… why, is the question.

Under what cases would this enable something that was not possible before, or, what would be made easier with this feature?

Reply:
This RFC allows using more liberal style of writing code
1)  values-driven style of programming
Instead of writing 
```rust
let a = Box::new(5);
let s = String::from("some string");
```
use
```rust
let a : *Box<u32> = 5;
let s : *String = "some string";

let a2 : Box<u32> =& 5;
let s2 : String =& "some string";
```
2) "On change update types, not logic" style of  programming
Let we have
```rust
let a : u32 = 5;
...;
let b = 3 + a;
```
But we decided  to change `a` to `Box<u32>`
So we update type and nothing more:
```rust
let a : *Box<u32> = 5;
```
Then we decided  to change `a` to `Rc<u32>`
So we update type and nothing more:
```rust
let a : *Rc<u32> = 5;
```

Reply:
That's… not compelling motivation. You're asking to add a very complicated feature to the type system because you would rather type `*Box<u32>` instead of `Box::new`?

And there are loads of unanswered questions here. What does this code even mean? Just saying "this value is magically inside a box now" doesn't put it there. How does the desugaring work?

And on that note, simply boxing something doesn't automatically make that meaningful. You mention `Rc` for example, but simply creating an `Rc` isn't enough to use it everywhere; unless you change other functions to also accept `Rc`, you're at best just copying the same integer directly out of the `Rc` immediately after you construct it, defeating the purpose of it.

Reply:
1) About desugaring. Ok, I'll add this into `Reference-level explanation`.
Actually primitive types
```rust
let a : u32 = 5;
let b : u32 = 4 + a;
```
Rust desugars into something like
```rust
// Stack<T> must implement  Construct<T> Trait
impl<T> Construct<T> for Stack<T> {
   fn construct(t : T) -> Stack<T> {
      Stack::new(t)
   }
}

let a : *Stack<u32> = Stack::constuct(5)::as_deref_type();
let b : *Stack<u32> = Stack::constuct(4 + a::deref())::as_deref_type();
```
which desugars further by ` Construct<T>` Trait into
```rust
let a : *Stack<u32> = Stack::new(5) as *Stack<u32>;
let b : *Stack<u32> = Stack::new(4 + a::deref()) as *Stack<u32>;
```
So, by analogy
```rust
let a : *Box<u32> = 5;
let s : *String = "some string";
let z : *Box<*String> = "some string";
```
must desugars into
```rust
let a : *Box<u32> = Box::constuct(5)::as_deref_type();
let s : *String = String::construct("some string")::as_deref_type();
let z : *Box<*String> = Box::constuct( String::construct("some string")::as_deref_type() )::as_deref_type();
```
which desugars further by ` Construct<T>` Trait into
```rust
let a : *Box<u32> = Box::new(5) as *Box<u32>;
let s : *String = String::from("some string") as *String;
let z : *Box<*String> = Box::new( String::from("some string") as *String ) as *Box<*String>;
```

2) About motivation: developers add `&str` type as a hack in the type system, but it is just an ordinary `str` Derefered Composite Type.
If DCT was allowed, it was not need special constructor (right vanilla syntax must be `let s : &str = &String::from( "sss")`), special dereferer (right vanilla syntax must be `let s12 : &str = &(s1.to_owned() + *s2)`)
DCT makes unimaginable behavior possible and easy to perform.
Other primitive types also use in dereferd form. 
Why did developers add additional syntax `let a : u32 = 5;` instead of use existed syntax already 
`let a : Stack<u32> = Stack::new(5);`?
I just propose some unification.

Reply:
> Rust desugars into something like
> 
> // Stack<T> must implement  Construct<T> Trait
> impl<T> Construct<T> for Stack<T> {
>    fn construct(t : T) -> Stack<T> {
>       Stack::new(t)
>    }
> }
> 
> let a : *Stack<u32> = Stack::constuct(5)::as_deref_type();
> let b : *Stack<u32> = Stack::constuct(4 + a::deref())::as_deref_type();

I've never heard of it. Any references?

Reply:
This desugaring doesn't exist by any definition of "desugaring". The local variable introduces a new place and the initializer gets stored into that place. In codegen, this place will then either be on the stack, in a register or completely optimized away.

Reply:
For example: [Storing Values with Variables](https://doc.rust-lang.org/stable/book/ch02-00-guessing-game-tutorial.html#storing-values-with-variables):

> `let apples = 5;`  
The equal sign (=) tells Rust we want **to bind** something to the variable now.

and this [Using Box<T> to Point to Data on the Heap](https://doc.rust-lang.org/stable/book/ch15-01-box.html#using-boxt-to-point-to-data-on-the-heap) : 

> Boxes allow you to store data on the heap **rather than the stack**. What remains on the stack is the pointer to the heap data.
Boxes don’t have performance overhead, other than storing their data on the heap instead of on the stack.

@lebensterben , you are right, I wrote not a code, but a pseudo-code, how compiler interpret instructions.

Reply:
Firstly, did you create this RFC because there is a desire for this feature from others? If so, please link a source.

> About motivation: developers add &str type as a hack in the type system, but it is just an ordinary str Derefered Composite Type.

References are a core aspect of Rust, and dynamically sized types such as `str`, while inconvenient at times, are far richer than just `str`. What about `&str` is a hack?

> If DCT was allowed, it was not need special constructor (right vanilla syntax must be let s : &str = &String::from( "sss")), special dereferer (right vanilla syntax must be let s12 : &str = &(s1.to_owned() + *s2))
DCT makes unimaginable behavior possible and easy to perform.

Constructors are just function calls. It seems you take issue with `Deref`/`DerefMut`, but your proposal adds complexity and doesn't seem to be any more expressive. It just reduces character count by a bit.

> Why did developers add additional syntax let a : u32 = 5; instead of use existed syntax already
let a : Stack<u32> = Stack::new(5);?

I'm not sure what you're saying with this contrast of "additional syntax" and "existed syntax". To be clear, we use `let a: u32 = 5;` and that's just as simple as it should be.

Reply:
> For example: Storing Values with Variables:
> > ...

> and this Using Box to Point to Data on the Heap
> > ...

No one is disagreeing with you on these sources, but nowhere do these sources mention `Stack<T>` or anything like that. You can look in the compiler or standard library source yourself; you won't find anything like this. If you're proposing that we add `Stack<T>` and `trait Construct`, that's entirely unnecessary. If you're just ultimately pushing for alternative syntax for `Box::new(4)` and whatnot, say it outright. I don't think the `=&` is reasonable though.

Reply:
@vacuus I proposed this RFC, because Rust type system do not forbid such types

Reply:
That's true, but that's not a use case.

Reply:
So for `let a: *Rc<String> = "foo";`, when you pass `a` directly to another function that accepts arbitrary type, is it expected to pass `Rc<String>` or `String` or what?

Reply:
@SOF3, that's a good question! This is a bit counterintuitive, but Derefered Types focus on moves, not borrowing (like primitive types). So:
```rust
let a :: *Rc<String>;
// and it is used inside function body as ( &a : &*Rc<String>);
let b = foo( &a : Rc<String>); 
let b = bar( a : String);

let c : *Rc<String>;
let d : *Rc<String> = c; // this is newly constructed *Rc<String>
let e /* : String */ = d;
```
Note: variable with Dereferd Type could be constructed if type is explicitly known!

Reply:
So this basically means we will start having implicit clones everywhere. Sounds like the start of performance footguns in Rust?

Reply:
@SOF3 You could create a footgun even with vanilla Rust. So yes, Derefered Types allow create footguns too, and yes, a bit easier. 
Bu it isn't necessary to have implicit clones. It is enough to have additional reference to have explicit "ordinary" Referenced Type:
```rust
let b = foo( &&a : &Rc<String>);
// and it is used inside function body as ( &a : &Rc<String>);
```
That why I'm thinking now about advantages of DT without its disadvantages: my [Smart constructors](https://github.com/rust-lang/rfcs/pull/3408#discussion_r1161094008) comment